src/analysis/locate.ml

```diff
@@ -410,59 +411,125 @@ let find_source ~config loc =
-              | 0 -> 0
-              | n -> (
-                (* Check if we are referring to the same files.
+          List.sort_uniq
+            ~cmp:(fun ((i : int), s) ((j : int), t) ->
+              let tmp = compare j i in
+              if tmp <> 0 then tmp
+              else
+                match compare s t with
+                | 0 -> 0
+                | n -> (
+                  (* Check if we are referring to the same files.
                     Especially useful on OSX case-insensitive FS.
                     FIXME: May be able handle symlinks and non-existing files,
                     CHECK *)
-                match (File_id.get s, File_id.get t) with
-                | s', t' when File_id.check s' t' -> 0
-                | _ -> n))
-          lst
-      in
-      match lst with
-      | (i1, _) :: (i2, _) :: _ when i1 = i2 -> Multiple_matches files
-      | (_, s) :: _ -> Found s
-      | _ -> assert false))
+                  match (File_id.get s, File_id.get t) with
+                  | s', t' when File_id.check s' t' -> 0
+                  | _ -> n))
+            lst
+        in
+        match lst with
+        | (i1, _) :: (i2, _) :: _ when i1 = i2 -> Multiple_matches files
+        | (_, s) :: _ -> Found s
+        | _ -> assert false)
+      | files_matching_digest ->
+        log ~title:"find_source" "... found multiple files with matching digest";
+        log ~title:"find_source"
+          "... using directory heuristic to choose the best one";
+        (* Give each source file a score that represents how close its path is to the
+           target path (the path of the build artifact) and then choose the source file
+           with the highest score.
+
+           The score of a source file is the longest tail of the path of the its
+           directory that is a subpath of the target path. This is premised on build
+           systems liking to put artifacts in paths that are similar to the source path.
+           i.e., dune may put the cmt for foo/bar/baz.ml in
+           _build/default/foo/bar/.bar.objs/byte/bar__Baz.cmt, so we want to use that
+           shared foo/bar in the path to disambiguate.
+
+           ex:
+             source file: /a/b/c/d/e/f.ml
+             target path: /a/b/c/_build/default/d/e/artifacts/f.cmi
+             score: 2, because /a/b/c/d/e is the source file's directory, and d/e is
+                    the longest tail of it that is a subpath of the target path. *)
+        let score_file source_file =
+          (* This is technically quadratic, but
+             a) most file paths are short
+             b) in the common case, this is linear because common_prefix_len
+                will usually fail on the first loop
+             c) this isn't a hot path - this is only for the uncommon case where there are
+                two identical files
+             So the stars would need to align for this to cause performance problems *)
```

Comment:
I think I agree with that. I was wondering whether a project-wide occurrences query returning a large number of results in a file that is concerned by this issue would cause a noticeable slowdown.

A possible mitigation would be to memoize the result, but it is probably not worth the trouble. It seems sufficiently unlikely to ignore it.